enhancements/sig-policy/85-gatekeeper-policy-integration/README.md

```diff
@@ -0,0 +1,257 @@
+objects can't own cluster scoped objects in Kubernetes, an alternative is required.
+
+The Template Sync controller would not put owner references on Gatekeeper `ConstraintTemplate`/constraints, but instead
+add a finalizer of `policy.open-cluster-management.io/cleanup-cluster-scoped-policies` to the `Policy` object that
+defines the Gatekeeper `ConstraintTemplate`/constraints. This finalizer would indicate that Gatekeeper
+`ConstraintTemplate`/constraints need to be removed when the OCM `Policy` object is deleted and if applicable, when the
+`Policy` object is updated. To be able to track which Gatekeeper constraints belong to a `Policy` object, a new label
+should be added to the Gatekeeper `ConstraintTemplate` objects and constraints of
+`policy.open-cluster-management.io/policy` with the policy name. Note that a finalizer will also need to be added to the
+`governance-policy-framework` `Deployment` like the Configuration Policy controller does to prevent the Template Sync
+from being deleted before the finalizer can be handled.
+
+The Template Sync controller would also need to generate a status event when it creates or updates Gatekeeper
+`ConstraintTemplate` objects since they are implementation details for the Gatekeeper constraints and don't provide any
+validation on their own.
+
+If the Gatekeeper `ConstraintTemplate` includes invalid rego syntax, a Gatekeeper validation webhook automatically
+rejects the requst as shown below. When this occurs, the OCM `Policy` will be set to non-compliant and include the error
+returned by the validation webhook.
+
+```bash
+$ kubectl apply -f bad-constraint-template.yaml
+Error from server (1 error occurred: templates["admission.k8s.gatekeeper.sh"]["K8sRequiredLabels"]:7: rego_parse_error: unexpected eof token
+          mis
+            ^): error when creating "temp6.yaml": admission webhook "validation.gatekeeper.sh" denied the request: 1 error occurred: templates["admission.k8s.gatekeeper.sh"]["K8sRequiredLabels"]:7: rego_parse_error: unexpected eof token
+          mis
+            ^
+```
+
+#### New Controller
+
+In the `governance-policy-framework-addon` controller manager, a new controller called `gatekeeper-constraint` would be
+added. This is a bit tricky because each Gatekeeper `ConstraintTemplate` produces a new custom resource definition (CRD)
+and thus the watches aren't known at controller instantiation time.
+
+To handle this, the controller would watch all `Policy` objects in the managed cluster namespace. During the reconcile,
+the controller will detect if Gatekeeper constraints are managed by the policy. If so, the
+[dynamic watch](https://github.com/stolostron/kubernetes-dependency-watches) library will be used to watch the
+individual Gatekeeper constraints and trigger reconciles on the `Policy` parent object whenever they have status
+updates.
+
+The new controller would also examine the `Policy` status before sending a status event to avoid sending duplicate
```

Comment:
Do other policy controllers do this? Should they?

Reply:
Other controllers don't currently and this is likely to avoid coupling policy controllers to how status is stored in the parent policy, however, I'm not opposed to this if there is a common library (e.g. policy-nucleus).

For this particular case, I think it makes sense because the policy enforcer (Gatekeeper) is not the one producing the policy framework status event, so there needs to be a way to know what status events have been relayed from Gatekeeper.